src/webgpu/api/validation/copyTextureToTexture.spec.ts

```diff
@@ -274,157 +281,311 @@ g.test('depth_stencil_copy_restrictions')
+          { x: 0, y: 0, width: -1, height: 0 },
+          { x: 0, y: 0, width: 0, height: -1 },
         ])
       )
       .combine(
-        poptions('dstCopyOrigin', [
-          { x: 0, y: 0, z: 0 },
-          { x: 1, y: 0, z: 0 },
-          { x: 0, y: 1, z: 0 },
-          { x: 1, y: 1, z: 0 },
-        ])
-      )
-      .combine(poptions('srcTextureSize', [64, 32]))
-      .combine(poptions('dstTextureSize', [64, 32]))
-      .combine(poptions('srcCopyLevel', [1, 2]))
-      .combine(poptions('dstCopyLevel', [0, 1]))
-      .expand(p =>
-        poptions('copyWidth', [
-          Math.max(p.srcTextureSize >> p.srcCopyLevel, p.dstTextureSize >> p.dstCopyLevel) -
-            Math.max(p.srcCopyOrigin.x, p.dstCopyOrigin.x),
-          8,
+        poptions('srcTextureSize', [
+          { width: 64, height: 64, depth: 1 },
+          { width: 64, height: 32, depth: 1 },
+          { width: 32, height: 32, depth: 1 },
         ])
       )
-      .expand(p =>
-        poptions('copyHeight', [
-          Math.max(p.srcTextureSize >> p.srcCopyLevel, p.dstTextureSize >> p.dstCopyLevel) -
-            Math.max(p.srcCopyOrigin.y, p.dstCopyOrigin.y),
-          8,
+      .combine(
+        poptions('dstTextureSize', [
+          { width: 64, height: 64, depth: 1 },
+          { width: 64, height: 32, depth: 1 },
+          { width: 32, height: 32, depth: 1 },
         ])
       )
+      .combine(poptions('srcCopyLevel', [1, 2]))
+      .combine(poptions('dstCopyLevel', [0, 1]))
   )
   .fn(async t => {
     const {
       format,
-      srcCopyOrigin,
-      dstCopyOrigin,
+      copyBoxOffsets,
       srcTextureSize,
       dstTextureSize,
       srcCopyLevel,
       dstCopyLevel,
-      copyWidth,
-      copyHeight,
     } = t.params;
 
     const kMipLevelCount = 3;
     const srcTexture = t.device.createTexture({
-      size: { width: srcTextureSize, height: srcTextureSize, depth: 1 },
+      size: { width: srcTextureSize.width, height: srcTextureSize.height, depth: 1 },
       format,
       mipLevelCount: kMipLevelCount,
       usage: GPUTextureUsage.COPY_SRC,
     });
     const dstTexture = t.device.createTexture({
-      size: { width: dstTextureSize, height: dstTextureSize, depth: 1 },
+      size: { width: dstTextureSize.width, height: dstTextureSize.height, depth: 1 },
       format,
       mipLevelCount: kMipLevelCount,
       usage: GPUTextureUsage.COPY_DST,
     });
 
+    const srcSizeAtLevel = t.GetPhysicalSubresourceSize(srcTextureSize, srcCopyLevel);
+    const dstSizeAtLevel = t.GetPhysicalSubresourceSize(dstTextureSize, dstCopyLevel);
+
+    const copyOrigin = { x: copyBoxOffsets.x, y: copyBoxOffsets.y, z: 0 };
+
+    const copyWidth =
+      Math.min(srcSizeAtLevel.width, dstSizeAtLevel.width) + copyBoxOffsets.width - copyOrigin.x;
+    const copyHeight =
+      Math.min(srcSizeAtLevel.height, dstSizeAtLevel.height) + copyBoxOffsets.height - copyOrigin.y;
+
     const isSuccess =
```

Comment:
Took me a bit to figure out why these are the conditions. Suggest:
```suggestion
    // Depth/stencil copies must copy whole subresources.
    const isSuccess =
```